### PR TITLE
Refactoring wagtail.contrib.redirects.views into a class based views

### PR DESCRIPTION
**Pull request by Temidayo32**
This pull request references issue #8365. This addresses the `wagtail/contrib/redirects/views.py`.  The following `views` were refactored:
- `wagtail/contrib/redirects/view.Index`
- `wagtail/contrib/redirects/view.Edit`
- `wagtail/contrib/redirects/view.Delete`
- `wagtail/contrib/redirects/view.Create`

### CODE REVIEW

**Comment by Temidayo32** [comment]
@laymonage  @lb-  @thibaudcolas  Please review. There are some failed test. I mentioned one in the #8365.

File: wagtail/contrib/redirects/templates/wagtailredirects/index.html [hunk 1]
```diff
@@ -1,10 +1,10 @@
-{% extends "wagtailadmin/base.html" %}
+{% extends "wagtailadmin/generic/base.html" %}
 {% load i18n %}
 {% load static wagtailadmin_tags %}
 
 {% block bodyclass %}redirects{% endblock %}
 
-{% block titletag %}{% trans "Redirects" %}{% endblock %}
+<!-- {% block titletag %}{% trans "Redirects" %}{% endblock %} -->
```
Comment (lb-) [inline wagtail/contrib/redirects/templates/wagtailredirects/index.html:7]: If we don't need this - we should remove it completely, not just comment it out.

**Review by lb- — COMMENTED** [review]
Just one comment for now, if you feel this is ready for a review, please mark as ready (not draft). Thanks for getting this prepared.

**Comment by lb-** [comment]
Some of the CI failures are just formatting, running `make format` should fix these.

**Comment by Temidayo32** [comment]
@lb- thanks so much for the response.

File: wagtail/contrib/redirects/templates/wagtailredirects/results.html [hunk 2]
```diff
@@ -2,7 +2,7 @@
 {% if redirects %}
     {% if query_string %}
         <h2 role="alert">
-            {% blocktrans trimmed count counter=redirects.paginator.count %}
+            {% blocktrans trimmed count counter=redirects.count %}
```
Comment (laymonage) [inline wagtail/contrib/redirects/templates/wagtailredirects/results.html:5]: This seems to cause the pagination to not show the page count correctly:

Page of . (missing the page numbers)

**Review by laymonage — CHANGES_REQUESTED** [review]
Thank you! This is a great start, although I think we could make more use of the generic views' templates as well.

File: wagtail/contrib/redirects/templates/wagtailredirects/results.html (hunk 2 above)
Comment (Temidayo32) [reply wagtail/contrib/redirects/templates/wagtailredirects/results.html:5]: Yeah, I will inspect the data sent from the view again for the attribute to use instead of `count`. After refactoring `paginator.count ` doesn't work anymore as it is an object. It only works for the FBV

File: wagtail/contrib/redirects/views.py [hunk 3]
```diff
@@ -35,178 +33,139 @@
...
-            return redirect("wagtailredirects:index")
+class Index(IndexView):
+    """
+    Lists all redirects for management within the admin.
+    """
+
+    template_name = "wagtailredirects/index.html"
+    results_template_name = "wagtailredirects/results.html"
+    any_permission_required = ["add", "change", "delete"]
+    permission_policy = ModelPermissionPolicy(Redirect)
```
[the remark this answers is not in the document]
Comment (Temidayo32) [reply wagtail/contrib/redirects/views.py:49]: Alright. Thanks.

File: wagtail/contrib/redirects/views.py [hunk 4]
```diff
@@ -35,178 +33,139 @@
...
+    form_class = RedirectForm
+    template_name = "wagtailredirects/add.html"
+    add_url_name = "wagtailredirects:add"
+    index_url_name = "wagtailredirects:index"
+    edit_url_name = "wagtailredirects:edit"
+    delete_url_name = "wagtailredirects:delete"
+    header_icon = "redirect"
+    page_title = gettext_lazy("Add redirect")
+    success_message = gettext_lazy("Redirect '%(object)s' created.")
+    error_message = "The redirect could not be created due to errors."
```
[the remark this answers is not in the document]
Comment (Temidayo32) [reply wagtail/contrib/redirects/views.py:168]: Alright. I will make the change. thanks.

File: wagtail/contrib/redirects/templates/wagtailredirects/confirm_delete.html [hunk 5]
```diff
@@ -7,7 +7,7 @@
 
     <div class="row row-flush nice-padding">
         <p>{% trans "Are you sure you want to delete this redirect?" %}</p>
-        <form action="{% url 'wagtailredirects:delete' redirect.id %}" method="POST">
+        <form action="{% url 'wagtailredirects:delete' redirect.pk %}" method="POST">
```
[the remark this answers is not in the document]
Comment (Temidayo32) [reply wagtail/contrib/redirects/templates/wagtailredirects/confirm_delete.html:10]: Oh, I see what you are saying. I will try that.

File: CHANGELOG.txt [hunk 6]
```diff
@@ -98,6 +98,10 @@ Changelog
  * Maintenance: Add changelog and issue tracker links to the PyPI project page (Panagiotis H.M. Issaris)
  * Maintenance: Add better deprecation warnings to the `search.Query` & `search.QueryDailyHits` model, move final set of templates from the admin search module to the search promotions contrib module (LB (Ben) Johnston)
  * Maintenance: Add generic `InspectView` to `ModelViewSet` (Sage Abdullah)
+ * Maintenance: Adopt `wagtail.admin.views.generic.IndexView` for Redirects index listing and search results (Temidayo Azeez)
+ * Maintenance: Adopt  `wagtail.views.generic.EditView for Redirects Edit View` (Temidayo Azeez)
+ *Maintenance: Adopt `wagtail.admin.views.generic.CreateView` for Redirects Create view (Temidayo Azeez)
+ * Maintenance: Adopt `wagtail.admin.views.generic.DeleteView` for Redirects Delete View (Temidayo Azeez)
```
[the remark this answers is not in the document]
Comment (Temidayo32) [reply CHANGELOG.txt]: resolved

**Comment by Temidayo32** [comment]
@laymonage So here is some feedback on using the Generic templates in place of the redirects templates.

Here is the Index view Generic Template (`wagtailadmin/generic/index.html`) below. As can be seen, the `import/export redirects` button is missing in this. However, the behaviour is as expected (the search works, so does the `add redirects` button)

DYdJpwLM6U

Here is the Index view Redirects Custom Template (`wagtailredirects/index.html`) below. The behaviour changes on refactoring to CBV. I haven't been able to find the cause of this. The issue here is, once a query is typed in the search, the panels disappear  and the `redirect` header gets duplicated, and an empty space appears on the left. Not sure what the cause of this template duplication is yet. 

8v04VTfxin


Here is the Generic Edit Template (`wagtailadmin/generic/edit.html`). As can be seen, this is a good replacement, and good to go without any more modifications. Except of course, if we would rather `delete redirect` over just `delete` text.

image


Here Redirect Edit Template (`wagtailredirects/edit.html`). It has a bit of unnecessary component displayed at the top. This is coming from the `wagtailadmin/shared/header.html`. 

image

For the Create Template, the difference between `wagtailredirects/add.html` and `wagtailadmin/generic/create.html` is that while the former uses `save` for its button, the latter uses `create` for its button. I used the former.

uTcZfu8727


The Delete Generic template works perfectly with no need for any modification, so I used it in place of the redirect delete template.

**Comment by lb-** [comment]
How's this going @Temidayo32 - are you up for continuing this? Looks like it needs a rebase on main sorry. 

Is there maybe a way to do a smaller scoped version of this PR without solving all the gaps we have in some shared templates?

**Comment by lb-** [comment]
Closing in favour of newer PR. 

https://github.com/wagtail/wagtail/pull/11762

**Comment by laymonage** [comment]
Hey @Temidayo32, apologies I didn't get to revisit this PR sooner. It's a bit hard to review all of the views at once unless they only require minor changes, as the suggestions would need to be applied to all of them, and repeating the same suggestions on the PR makes it hard to follow. If you or anyone else is interested in picking this up, I gave a bit more details here: https://github.com/wagtail/wagtail/issues/8365#issuecomment-2061032553

I realised I haven't added you as the co-author for #11762. Will fix that soon. Thanks again!
